kv/txn.go

```diff
@@ -17,23 +17,112 @@ package kv
 	"github.com/pingcap/tidb/util/logutil"
+	"github.com/tikv/client-go/v2/oracle"
 	"go.uber.org/zap"
 )
 
+const (
+	// TimeToPrintLongTimeInternalTxn is the duration if the internal transaction lasts more than it,
+	// TiDB prints a log message
+	TimeToPrintLongTimeInternalTxn = time.Minute * 5
+)
+
+var globalInnerTxnTsBox = innerTxnStartTsBox{
+	innerTSLock:        sync.Mutex{},
+	innerTxnStartTsMap: make(map[uint64]struct{}, 256),
+}
+
+type innerTxnStartTsBox struct {
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]struct{}
+}
+
+// wrapStoreInterTxnTS is the entry function used to store the startTS of an internal transaction to globalInnerTxnTsBox
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapStoreInterTxnTS(startTS uint64) {
+	globalInnerTxnTsBox.storeInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) storeInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	ib.innerTxnStartTsMap[startTS] = struct{}{}
+	ib.innerTSLock.Unlock()
+
+}
+
+// wrapDeleteInterTxnTS delete the startTS from globalInnerTxnTsBox when the transaciotn is commted
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapDeleteInterTxnTS(startTS uint64) {
+	globalInnerTxnTsBox.deleteInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) deleteInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	delete(ib.innerTxnStartTsMap, startTS)
+	ib.innerTSLock.Unlock()
+}
+
+// WrapGetMinStartTs get the min StartTS between startTSLowerLimit and curMinStartTS in globalInnerTxnTsBox
+func WrapGetMinStartTs(now time.Time, startTSLowerLimit uint64,
+	curMinStartTS uint64) uint64 {
+	return globalInnerTxnTsBox.getMinStartTs(now, startTSLowerLimit, curMinStartTS)
+}
+
+func (ib *innerTxnStartTsBox) getMinStartTs(now time.Time, startTSLowerLimit uint64,
+	curMinStartTS uint64) uint64 {
+	minStartTS := curMinStartTS
+	ib.innerTSLock.Lock()
+	for innerTS := range ib.innerTxnStartTsMap {
+		PrintLongTimeInternalTxn(now, innerTS, true)
+		if innerTS > startTSLowerLimit && innerTS < minStartTS {
+			minStartTS = innerTS
+		}
+	}
+	ib.innerTSLock.Unlock()
+	return minStartTS
+}
+
+// PrintLongTimeInternalTxn print the internal transaction information
+// runByFunction	true means the transaction is run by `RunInNewTxn`,
+//					false means the transaction is run by internal session
+func PrintLongTimeInternalTxn(now time.Time, startTS uint64, runByFunction bool) {
+	if startTS > 0 {
+		innerTxnStartTime := oracle.GetTimeFromTS(startTS)
+		if now.Sub(innerTxnStartTime) > TimeToPrintLongTimeInternalTxn {
+			callerName := "internal session"
+			if runByFunction {
+				callerName = "RunInNewTxn"
+			}
+			infoHeader := fmt.Sprintf("An internal transaction running by %s lasts long time", callerName)
+
+			logutil.BgLogger().Info(infoHeder,
+				zap.Duration("time", now.Sub(innerTxnStartTime)), zap.Uint64("startTS", startTS),
+				zap.Time("start time", innerTxnStartTime))
```

Comment:
The transaction fingerprint may help to output more information about what actually is the inner transaction doing.
The `start_ts` itself could not help much if there's not enough logs about it, we could do some improvements in the future
/cc @longfangsong